### PR TITLE
Add Stream.iterate APIs that takes predicate

### PR DESCRIPTION
**Pull request by treblereel**

### CODE REVIEW

File: jre/java/java/util/stream/DoubleStream.java [hunk 1]
```diff
@@ -147,19 +147,35 @@ public boolean tryAdvance(DoubleConsumer action) {
...
           private double next = seed;
+          private boolean terminated = false;
 
           @Override
           public boolean tryAdvance(DoubleConsumer action) {
+            if (terminated) {
+              return false;
+            }
+            if (!first) {
+              next = f.applyAsDouble(next);
```
Comment (gkdn) [inline jre/java/java/util/stream/DoubleStream.java:167]: Looks like there is behavior change. In the original logic we were calling apply even for the first item. We are no longer doing that. Looking at the test, that seems intentional...

File: jre/javatests/com/google/j2cl/jre/java8/util/stream/DoubleStreamTest.java [hunk 2]
```diff
@@ -116,6 +116,22 @@ public void testIterate() {
...
+    int[] calledCount = {0};
+    double[] array = DoubleStream.iterate(0, val -> {
+      calledCount[0]++;
+      return val + 1;
+    }).limit(5).toArray();
+    // Verify that the function was called for each value after the seed
+    assertEquals(array.length - 1, calledCount[0]);
+    // Sanity check the values returned
+    assertEquals(new double[] {0, 1, 2, 3, 4}, array);
   }
```
Comment (gkdn) [inline jre/javatests/com/google/j2cl/jre/java8/util/stream/DoubleStreamTest.java:135]: Could we add test to cover use of predicate?

File: jre/java/java/util/stream/DoubleStream.java [hunk 3]
```diff
@@ -147,19 +147,35 @@ public boolean tryAdvance(DoubleConsumer action) {
...
 
   static DoubleStream iterate(double seed, DoubleUnaryOperator f) {
+    return iterate(seed, ignore -> true, f);
+  }
+
+  static DoubleStream iterate(double seed, DoublePredicate hasNext, DoubleUnaryOperator f) {
     Spliterator.OfDouble spliterator =
         new Spliterators.AbstractDoubleSpliterator(
             Long.MAX_VALUE, Spliterator.IMMUTABLE | Spliterator.ORDERED) {
+          private boolean first = true;
```
[the remark this answers is not in the document]
Comment (treblereel) [reply jre/java/java/util/stream/DoubleStream.java]: @gkdn, My fault, I have forgot to add it to the commit

File: jre/javatests/com/google/j2cl/jre/java8/util/stream/DoubleStreamTest.java (hunk 2 above)
Comment (treblereel) [reply jre/javatests/com/google/j2cl/jre/java8/util/stream/DoubleStreamTest.java:154]: @gkdn yeap, I have added the tests

File: jre/java/java/util/stream/DoubleStream.java (hunk 1 above)
Comment (treblereel) [reply jre/java/java/util/stream/DoubleStream.java:167]: @gkdn I have added tests that check this case

**Pull request closed without merging**